### ISSUE TITLE
[2020] Don't Stop Pretraining: Adapt Language Models to Domains and Tasks

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#53]
## 論文リンク
https://arxiv.org/abs/2004.10964

## 公開日（yyyy/mm/dd）
2020/04/23

## 概要
あとで書く。

**Comment by yoheikikuta** [comment]
ACL で面白そうな論文ないかな〜と思って探した中で見つけた第二弾。
pre-training が NLP タスクでも広く使われるようになって久しいが、domain specific な状況で使う時の学習に関して何か新しい知見があるかなと思って読んでみることにした。

ちなみにこの論文は `Honorable Mention Papers – Main Conference` に選ばれている。

**Comment by yoheikikuta** [comment]
BERT あたりから一気に NLP の pre-training も進んだが、その後継の ROBERTA では圧縮なしのテキスト 160[GB] で pre-training していて、そこで得られた分散表現は様々な source で様々なデータサイズで様々なタスクに高い性能を発揮している。

大体のタスクだともうこういう pre-trained モデル使えばええんや！となりそうだが、特定のドメインに特化させた別個の pre-trained model がまだ有効なのか？というのは十分には明らかになっていない。

例えば https://arxiv.org/abs/1901.08746 の論文では BERT を生物医学系の corpus に特化させたもので、PubMed などの domain specific なデータで pre-training してさらに

**Comment by yoheikikuta** [comment]
2018 以降、NLP のモデルはまず言語モデル（BERT 系譜なら Masked Language Model）として教師なし学習をして、その後 downstream タスクのために教師ありで fine-tuning する、という二段構えで学習される。

この論文では ROBERTA をベースラインに使う。
ROBERTA は大量のデータで学習してるが、それでも英語で現れる様々なバリエーションを一般化するほどの多様性があるかは不明（というよりきっとないだろうと思っている）なので、この論文では BIOMED, CS, NEWS, REVIEWS (amazon) の 4 つの domain データを使って DAPT や TAPT を試す。

**Comment by yoheikikuta** [comment]
実験に入る前に、domain 間の類似度に関して定量的に評価しておく。
とはいえこれはかなりシンプルな定義で、各 domain の corpus の中で、uniqram の上位 10K 個（ストップワードを除く）の overlap を測ったというものになっている。結果は以下。

<p align="center">
  <img src="https://i.imgur.com/8HDw5HF.png">
</p>

PT は ROBERTA の pre-training で使ったデータセットで、これは wikipedia などが含まれていて News データセットとの類似度が高いことが分かる。10K まで見たらそれぞれの類似度はもっと高くなるかなと思ったけど、思ったより overlap 低いなと思った。domain 的に結構専門用語とかも多そうなものがあるしこんなもんなのかね。

**Comment by yoheikikuta** [comment]
ということで実験。まずは DAPT の方から。

- データ前処理
  生物医学と CS の要約と本文の段落は、scispaCy を用いて文を分割して使用。各ニュース記事の要約と全文、Amazonレビューはレビュー本文を spaCy を用いて文の分割を実施して使用。

-DAPT
  pre-training の corpus は Table 1 で貼った量を使って、12.5K steps で学習している。corpus のサイズが違うが、これは token の数が違うことからも想像できるように、分の長さが結構違う。この corpus のサイズは学習時に同程度の文章量だけモデルに投入するように調整されたものになっている。以下で説明する classification task の domain と同じ domain で学習するのが DAPT である。

- classification task
  それぞれの domain で二つのタスクが利用可能になっている。詳細は以下。
  この段階ではまだよく分からんもの（high-resource settings はなんか使い方違うんだっけ、とか、Unl. は unlabeled っぽいけどこのデータも使うんだっけ、とか）もあるが、基本的には domain 毎の classification 用のデータですということ。

<p align="center">
  <img src="https://i.imgur.com/UYn7jyX.png">
</p>

- ¬DAPT
  これは異なる domain で学習するケースで、NEWS task -> CS pre-training, REVIEWS -> BIOMED, CS -> NEWS, BIOMED -> REVIEWS で試している。これは overlap の統計情報を参考にして決めている。

結果は以下のようになる。
domain を適切に合わせて pre-training した DAPT は良い性能を示している。一方で異なる domain データで pre-training をした ¬DAPT は多くの場合悪影響がある。追加で学習された domain に影響を受けすぎて本来解きたいタスクの domain に適応できてないことが伺える。一方でたまに素の ROBERTA よりよくなることもあったりして、まあ関係が薄いデータでも pre-training 続けるといいことあることもあると（理由がなぜとか全然分からんし、まあこんなこともあるよねくらいでとりあえず無視しておく）。

<p align="center">
  <img src="https://i.imgur.com/BQADdB4.png">
</p>

異なる randomly selected seed で 5 回ずつ実験していて、右下の数字が 1σ のズレである。
真面目に考えるとタスクによっては全然有意さあるとは言えないじゃんというものも出てくるが、こういう統計誤差が書いてないものも機械学習の論文では山ほどあるので、みんなちゃんと書いていってくれるようになるといいね（実験をする際にはマシンパワーの依存度が大きくなるという観点もあるのだが...）。

**Comment by yoheikikuta** [comment]
次は TAPT について。
task に使えるデータは少ないが、これはこれが属する domain の部分集合であるのでこいつを unlabeled pre-training に使うことによって性能が向上するだろうという戦略である。データ量が少ないけど task-relevant なので、うまくいくなら学習コストは抑えつつ精度を引き上げられる。

- TAPT
  task-specific なデータだけを使って pre-training をする。データ量が少ないので 100 epochs のみの学習。
  学習時に masking probaility 0.15 で epoch 毎に異なる mask をして augmentation してるって言ってるけど、ROBERTA は dynamic masking で pre-training するからこれは別に DAPT でもそうしてたと思うんだよな。なぜこっちだけ強調して書いてるのかは分からん。

結果は以下のテーブルのようになる。

<p align="center">
  <img src="https://i.imgur.com/3CDswxX.png">
</p>

あとで書く。

**Comment by yoheikikuta** [comment]
TAPT が使用データ量の割に良い結果を出してるので、これをもうちょっと効率的にやりたいという欲が出てくる。
解きたいタスクのデータに近いデータを持ってくるということをしたいわけだ。

論文の進め方としては以下のようになっている。
- タスクに近いデータと分かっているものを使うことで確かに性能が上がることを確かめる
- タスクに近いデータが分からない（実用上はこちらが多い）場合に近いデータと思われるものを検出する手法を考案・適用してみてその効果を確かめる

**Comment by yoheikikuta** [comment]
最初の方は Human Curated-TAPT と呼んでいる。
Human Curated と出てくるので「おっ、人力でデータ選別するのか？」と思ったが、これはタスクを作る時にデータソースが決まっていて、annotation cost の観点からその中の一部だけに label が付与されたというデータを使うという話だった。unlabled data も同じデータソースなので label がついたデータと分布は似ているだろうという仮定。

細かいセットアップは置いておいて、結果が以下の表。

**Comment by yoheikikuta** [comment]
curated TAPT がうまくいってそうなので、次はデータがたくさんあるけどどれがタスク分布に近いデータかが分からない時にそれを抽出する方法を考案して試す。

あとで書く。

**Comment by yoheikikuta** [comment]
What is a domain? というセクションがあってオッと思うが、これはみんななんとなくの意味で使っててちゃんとこのことについて考えてる研究は少ないぞって言ってる。
この論文では DAPT と TAPT が補完的であることから、様々な粒度でスペクトルを成してる（ `(e.g., Amazon reviews for a specific product, all Amazon reviews, all reviews on the web, the web)` ）んじゃないですかと投げかけてる。とりあえず言っときます的な書き方だね。

**Comment by yoheikikuta** [comment]
ということで一通り読んだ。
ACL の `Honorable Mention Papers – Main Conference` に選ばれたので面白いのかと思ったけど、pre-training について DAPT と TAPT を明確に区別していくつかの domain でそれをきっちり実験しましたという堅実な実験論文という感じだった。

データが少ない domain では TAPT のような方法が有効そう、というのは実戦でも使えそうだね。
